pkg/controller/util.go

```diff
@@ -1171,6 +1176,30 @@ func inflateSizeWithOverhead(c client.Client, imgSize int64, pvcSpec *v1.Persist
 	return returnSize, nil
 }
 
+// handleFailedPod handles pod-creation errors and updates the pod's PVC without providing sensitive information
+func handleFailedPod(err error, podName string, pvc *v1.PersistentVolumeClaim, recorder record.EventRecorder, c client.Client) error {
+	if err == nil {
+		return nil
+	}
+	// Generic reason and msg to avoid providing sensitive information
+	reason := ErrStartingPod
+	msg := fmt.Sprintf(MessageErrStartingPod, podName)
+
+	// Error handling to fine-tune the event with pertinent info
+	if errQuotaExceeded(err) {
+		reason = ErrExceededQuota
+	} // TODO: Add more error cases
+
+	recorder.Event(pvc, v1.EventTypeWarning, reason, msg)
+
+	AddAnnotation(pvc, AnnPodPhase, string(v1.PodFailed))
```

Comment:
DataVolumes should never be marked failed. That would violate the eventually consistency of datavolumes. We should emit an event (which we do here) and update the datavolume running condition to false with the reason why. It is entirely possible that the namespace this is running in will get more quota and the DV will succeed when that happens. If we mark it failed, that is it, the DV will no longer do anything and you have to delete it, and re-create it.

You can find functions to manipulate the conditions in datavolume-conditions.go

Reply:
I have some doubts regarding how to handle this. The problem is that the error handling function is mainly called outside the datavolume-controller, so I can't directly access the datavolume to modify its running condition.

I see two options regarding how to handle this in `handleFailedPod`, but both have their downsides:

- To add the `PodFailed` annotation in the PVC and then handle that in the datavolume-controller, so we change the **running condition** to false instead of the **phase** to failed. I see this option more in concordance with what you said in the comment, but I fear it might change some other behavior in the controller. **I've just implemented this one**.

- To change the running condition of the PVC to false, so the DV is updated in the datavolume-controller without modifying a lot of code. This option is way less intrusive, but I see a major problem: When either the clone or the size-detection pod fails, the PVC that should be updated is the source PVC, not the target one. This means that the running condition of the source DV will also change, which is unwanted behavior.

Reply:
Why would we update the source PVC? the source is completed, we have determined that. So we always update the target PVC. In fact we already have some handling on the bound condition in case the source PVC is not bound (which is also possible). Always update the target PVC.

Reply:
Since the size-detection pod and the source-clone pod both use the source PVC, I thought it made sense to change the `AnnPodFailed` annotation in the source and not the target in these cases. I agree that it makes more sense to change the running condition in the target PVC, but I still find it odd to handle a PVC that has no direct relation with the pod in `handleFailedPod`.

If it's preferred to use the target in all cases, I guess I can just change the running condition and let the datavolume-controller handle everything (second option). However, don't my changes regarding what to do with the `AnnPodFailed` annotation in the datavolume-controller make sense? What case would justify setting the DV to failed after a pod fails?